### PR TITLE
[gguf] Add types

### PR DESCRIPTION
**Pull request by mishig25**
GGUF add types. Follow up to https://github.com/huggingface/huggingface.js/pull/540#issuecomment-2000249807.

No any kind of validation, just types

cc: @biw also

### CODE REVIEW

File: packages/gguf/src/gguf.ts [hunk 1]
```diff
@@ -1,7 +1,9 @@
-export type MetadataBaseValue = string | number | bigint | boolean;
-export type MetadataValue = MetadataBaseValue | MetadataBaseValue[] | MetadataValue[]; /// recursive as arrays can be nested.
+import type { MetadataValue, Version, GGUFMetadata, GGUFTensorInfo, GGUFParseOutput } from "./types";
+import { GGUFValueType } from "./types";
+
+export type { MetadataValue, Version, GGUFMetadata, GGUFTensorInfo, GGUFParseOutput } from "./types";
+export { GGUFValueType, GGMLQuantizationType } from "./types";
```
Comment (mishig25) [inline packages/gguf/src/gguf.ts:5]: is it a correct way to re-export ?

File: packages/gguf/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	Q8_K = 15,
+	IQ2_XXS = 16,
+	IQ2_XS = 17,
+	IQ3_XXS = 18,
+	IQ1_S = 19,
+	IQ4_NL = 20,
+	IQ3_S = 21,
+	IQ2_S = 22,
+	IQ4_XS = 23,
+}
```
Comment (julien-c) [inline packages/gguf/src/types.ts:29]: enums are not strictly speaking types as they expose objects in the runtime.

End of not super useful pedantic note haha. cc @coyotte508

File: packages/gguf/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+type TransformerLLMArchitecture = Exclude<Architecture, "rwkv" | "whisper">;
+type TransformerLLM = ModelBase<TransformerLLMArchitecture> &
+	MOE<TransformerLLMArchitecture> &
+	Attention<TransformerLLMArchitecture> &
+	Rope<TransformerLLMArchitecture>;
+
+export type RWKV = ModelBase<"rwkv"> & { "rwkv.architecture_version": number };
+export type LLM = TransformerLLM | RWKV;
+export type Whisper = ModelBase<"encoder.whisper"> & ModelBase<"decoder.whisper">;
+export type Model = (LLM | Whisper) & Partial<Tokenizer>;
```
Comment (julien-c) [inline packages/gguf/src/types.ts:120]: very neat types (though they make my head hurt a bit, lol)

**Review by julien-c — APPROVED** [review]

File: packages/gguf/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	"llama",
+	"mpt",
+	"gptneox",
+	"gptj",
+	"gpt2",
+	"bloom",
+	"falcon",
+	"gemma",
+	"rwkv",
+	"whisper",
```
Comment (julien-c) [inline packages/gguf/src/types.ts:57]:
```suggestion
    "llama",
    "falcon",
    "baichuan",
    "gpt2",
    "gptj",
    "gptneox",
    "mpt",
    "starcoder",
    "persimmon",
    "refact",
    "bert",
    "nomic-bert",
    "bloom",
    "stablelm",
    "qwen",
    "qwen2",
    "phi2",
    "plamo",
    "codeshell",
    "orion",
    "internlm2",
    "minicpm",
    "gemma",
    "starcoder2",
    "mamba",
```

Comment (julien-c) [reply packages/gguf/src/types.ts:57]: (optional, but it's the current list from the llama.cpp source of truth IIUC)

Comment (mishig25) [reply packages/gguf/src/types.ts:57]: great ! Will update them and their specific types from (ex: https://github.com/ggerganov/ggml/blob/master/docs/gguf.md#mamba)

**Review by julien-c — APPROVED** [review]
i would merge as is and iterate later

**Pull request merged**